llvm/lib/Transforms/Coroutines/CoroFrame.cpp

```diff
@@ -3046,7 +3046,8 @@ void coro::buildCoroutineFrame(
   // Collect the spills for arguments and other not-materializable values.
   for (Argument &A : F.args())
     for (User *U : A.users())
-      if (Checker.isDefinitionAcrossSuspend(A, U))
+      if (Checker.isDefinitionAcrossSuspend(A, U) ||
+          isa<CoroEndInst>(U))
```

Comment:
We do we need to include CoroEnd explicitly here?

Reply:
The BBs with `coro.end` calls are a bit special during the coroutine frame calculation: they do not kill values (see https://github.com/llvm/llvm-project/blob/f8b88a838cde67c14c28d7b40efa9af842d010d8/llvm/lib/Transforms/Coroutines/CoroFrame.cpp#L270). As a result, values that have only a single `coro.end` use (e.g. input arguments directly used by `coro.end`) will not be saved in the frame as `isDefinitionAcrossSuspend` will return `false` for them.

Reply:
Got it. Then it smells slightly bad. We have a lot `isDefinitionAcrossSuspend` uses and it looks easy to forget similar cases. I feel better to solve the issue fundamentally by introducing new intrinsics.

e.g., we can introduce  `token  llvm.coro.end.retcont.return_values(...)` and refactor `llvm.coro.end()` to `declare i1 @llvm.coro.end(ptr <handle>, i1 <unwind>, token <return>)`. And for non-retcon-once ABI, we can fill the third argument as `none` simply. For retcon-once ABI, it should be easy to generate the actual ret too. 

Then we can avoid the `isDefinitionAcrossSuspend` problem automatically since the BB will be split around `coro.end`.

Reply:
Ok, I'll give it a try. Note that this will be a C++ API breaking change though as we will always need to pass an extra argument.

Reply:
Thanks. The API breaking is relatively acceptable since generally we don't have requirement for backport compatibility.